app/src/main/java/io/homeassistant/android/webview/WebViewActivity.kt

```diff
@@ -0,0 +1,108 @@
+package io.homeassistant.android.webview
+
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import android.os.Bundle
+import android.util.Log
+import android.webkit.JavascriptInterface
+import android.webkit.WebView
+import android.webkit.WebViewClient
+import androidx.appcompat.app.AppCompatActivity
+import io.homeassistant.android.BuildConfig
+import io.homeassistant.android.R
+import io.homeassistant.android.about.SettingsActivity
+import io.homeassistant.android.api.Session
+import io.homeassistant.android.io.homeassistant.android.api.Token
+import org.json.JSONObject
+
+
+class WebViewActivity : AppCompatActivity(), io.homeassistant.android.webview.WebView {
+
+    companion object {
+        private const val TAG = "WebviewActivity"
+
+        fun newInstance(context: Context): Intent {
+            return Intent(context, WebViewActivity::class.java)
+        }
+    }
+
+    private lateinit var webView: WebView
+    private lateinit var presenter: WebViewPresenter
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_webview)
+
+        presenter = WebViewPresenterImpl(this)
+
+        if (BuildConfig.DEBUG) {
+            WebView.setWebContentsDebuggingEnabled(true)
+        }
+
+        webView = findViewById(R.id.webview)
+        webView.apply {
+            settings.javaScriptEnabled = true
+            settings.domStorageEnabled = true
+            webViewClient = WebViewClient()
+        }
+
+        presenter.onViewReady()
+    }
+
+    override fun onBackPressed() {
+        if (webView.canGoBack()) {
+            webView.goBack()
+        } else {
+            super.onBackPressed()
+        }
+    }
+
+    override fun setupJavascriptInterface(token: Token) {
+        webView.addJavascriptInterface(object : Any() {
+            @JavascriptInterface
+            fun getExternalAuth(callback: String) {
+                webView.post {
+                    webView.evaluateJavascript(
+                        "${JSONObject(callback).get("callback")}(true, {\n" +
+                                "  \"access_token\": \"${token.accessToken}\",\n" +
```

Comment:
This seems like it is hardcoding the tokens from the first launch. However, if the token is expired in the frontend, this code should refresh the token before giving it to the app.

Reply:
I noticed token is refreshed if necessary when we open the webview. This will work ok if the user is not planning to use the app longer than that the token is currently valid. Better if we refresh it in this function, if necessary. 

Ideally, it would work like this:
 - the moment the app starts up and it has tokens, start a refresh request if they are no longer valid
 - in the JavaScript interface, if a refresh request is in progress (because app got opened), wait for that request to finish, otherwise start own request for refresh.

Reply:
To make it quicker I refresh the token (if needed) each time getExternalAuth is call. Is that ok for you ?